File: 13. 0-1 Knapsack Problem.cpp
// 13. 0/1 Knapsack Problem
// Topic: 0/1 Knapsack
// Description: Given weights and values, pick items to maximize value without exceeding capacity.
// Use memoization on (idx, remainingWeight). Input: n, capacity, then n pairs weight value Output:
// maximum value

#include <bits/stdc++.h>
using namespace std;
using ll = long long;

ll knap(int idx, int W, const vector<int>& wt, const vector<int>& val, vector<vector<ll>>& memo)
{
    if (idx >= (int)wt.size() || W <= 0)
        return 0;
    if (memo[idx][W] != -1)
        return memo[idx][W];
    ll skip = knap(idx + 1, W, wt, val, memo);
    ll take = 0;
    if (wt[idx] <= W)
        take = val[idx] + knap(idx + 1, W - wt[idx], wt, val, memo);
    memo[idx][W] = max(skip, take);
    return memo[idx][W];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, W;
    if (!(cin >> n >> W))
        return 0;
    vector<int> wt(n), val(n);
    for (int i = 0; i < n; ++i) cin >> wt[i] >> val[i];
    vector<vector<ll>> memo(n, vector<ll>(W + 1, -1));
    cout << knap(0, W, wt, val, memo) << '\n';
}

// Memoization process started in knap where memo[idx][W] caches optimal values for subproblems.
// Time Complexity: O(n * W)
// Space Complexity: O(n * W)

File: 13. 0-1 Knapsack Problem.py
# 13. 0/1 Knapsack Problem
# Topic: 0/1 Knapsack
# Description: Given weights and values, pick items to maximize value without exceeding capacity using recursion + memoization.
# Input: n W followed by n pairs (weight value)
# Output: maximum achievable value

import sys
from functools import lru_cache

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    n = int(next(it))
    W = int(next(it))
    wt = [0]*n
    val = [0]*n
    for i in range(n):
        wt[i] = int(next(it)); val[i] = int(next(it))

    @lru_cache(maxsize=None)
    def knap(i: int, remW: int) -> int:
        if i >= n or remW <= 0:
            return 0
        res = knap(i+1, remW)
        if wt[i] <= remW:
            res = max(res, val[i] + knap(i+1, remW - wt[i]))
        return res

    print(knap(0, W))

# Memoization process started at knap decorated with lru_cache.
# Time Complexity: O(n * W)
# Space Complexity: O(n * W)